Week2/ikasty

ikasty opened the pull request:
### group anagram
총 세 가지 버전으로 해결함
- 4d5f00f: 생각나는 대로 아무렇게나 짠 망한 버전
- 0843cd8: Trie 라는 자료구조를 사용하는, 나름 빠르지만 메모리 사용량이 미친듯이 올라가는 버전
- 8740db1: 해시를 쓰라는 힌트를 받고 아이디어를 떠올려서 최적화한 최종 버전

### 01 matrix 
풀기는 풀었으되 컨닝이 포함되어 있음
- 0923edd: 원래 아이디어대로 푼, 큐에다가 넣어가며 푼 버전
- b962977: 다 풀고난다음 성능이 억울해서 다른사람 답변 구경했다가, 스포일러 당하고 답변 개조해서 만든 `100.00%` 버전

ikasty committed:
1st try of group-anagrams

ikasty committed:
2nd version of group anagrams

ikasty committed:
Faster version of group anagrams

ikasty committed:
1st try of 01-matrix

ikasty committed:
add cheat version of 01-matrix

puchidev commented on challenges/week2.01-matrix/ikasty.scala:
```diff
@@ -0,0 +1,42 @@
+            val candidate = List((tuple._1 - 1, tuple._2), (tuple._1 + 1, tuple._2),
+                                 (tuple._1, tuple._2 - 1), (tuple._1, tuple._2 + 1))
+                // matrix 범위 내에 있으며
+               .filter(item => item._1 >= 0 && item._2 >= 0)
+               .filter(item => item._1 < matrix.size && item._2 < matrix.head.size)
+                // 방문하지 않은 노드이며
+               .filter(item => result(item._1)(item._2) < 0)
+                // 큐에 넣지 않은 노드라면
+               .filter(item => !queue.contains(item))
```
`candidate.filter()` 파트는 filter를 여러번 호출하든 아니면 로직을 병합해 적은 횟수를 호출하든 상관없이 비슷한 퍼포먼스를 내는 걸까요

ikasty replied:
스칼라는 실행 시 자바 바이트코드로 변환되어 실행되므로 아마도 이론상의 퍼포먼스는 동일할 것 같습니다. 실제 실험 결과도 그랬구요. 메모리 사용량에는 소폭 변화가 있는 것으로 봐선 완전히 동일한 건 아닌 것 같습니다.

puchidev commented on challenges/week2.01-matrix/ikasty.scala:
```diff
@@ -0,0 +1,42 @@
+               .filter(item => !queue.contains(item))
+
+            // 현재 result에서 1 증가시키고 큐에 넣기
+            candidate map { item => result(item._1)(item._2) = result(tuple._1)(tuple._2) + 1 }
```
이 부분은 map 대신 forEach로 쓸수 있는 것이 아닌가 하고 조심스레 질문해 봅니다😐🤐

ikasty replied:
그렇네여. 저도 스칼라를 어제 배워서 오늘 써먹는 수준이라...
마침 오늘 '함수의 부가 효과를 일으키는 반복문에선 forEach를 사용한다' 고 배웠습니다ㅎㅎㅎ

puchidev commented on challenges/week2.group-anagrams/ikasty.py:
```diff
@@ -0,0 +1,9 @@
+        for item in strs:
+            result[''.join(sorted(item))].append(item)
+
+        return result.values()
```
말이 필요없는 깔끔한 구현 + 나이스 퍼포먼스! 😎 파이썬은 역시 멋져

puchidev commented on challenges/week2.01-matrix/ikasty.cheat.scala:
```diff
@@ -0,0 +1,26 @@
+// Runtime: 684 ms, faster than 100.00% of Scala online submissions
+// Memory Usage: 60.3 MB, less than 91.67% of Scala online submissions
```
😵🤪😱🥳

puchidev commented on challenges/week2.01-matrix/ikasty.cheat.scala:
```diff
@@ -0,0 +1,26 @@
+        
+        result
+    }
+}
```
DP의 길은 정말 어렵네여 이런 솔루션을 필요할 때 바로 머릿속에서 고안한다? 흠터레스팅.. 저는 자신이 없군요

ikasty replied:
멀고도 험난한것...ㅠㅠ

puchidev reviewed (commented):
스칼라 잘 구경했어요 꿀잼🤣 구문에 놀라고 그럭저럭 이해가 되는 점에 놀랐구요. 짧은 질문 몇개 넣어봤으니 코멘트해 주시면 확인하고 리뷰승인 갈게요 😁

ikasty committed:
Apply reviews

https://github.com/atelier-algorithm/weekly-algos/pull/4#pullrequestreview-598826103

puchidev reviewed (approved):
스칼라 마술 잘 보고갑니다 또 재밌는거 기대해여! 😎
